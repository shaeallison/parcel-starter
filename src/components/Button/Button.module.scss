.button {
  background-color: rgb(203, 9, 103);
  color: var(--white from global);
  border: none;
  appearance: none;
  padding: 1rem 2rem;
  font-weight: 600;
  letter-spacing: 0.03rem;
  transition: 0.3s ease all;

  &:hover {
    background-color: rgb(178, 7, 90);
  }
}

.message {
  margin-top: 2rem;
}
