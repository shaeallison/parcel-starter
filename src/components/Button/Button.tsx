import { ReactNode, useState } from 'react'
import * as styles from './Button.module.scss'

type ButtonProps = {
  children: ReactNode
  id: string
}

export const Button = (props: ButtonProps): JSX.Element => {
  const [clicked, setClicked] = useState<boolean>(false)
  const { children, id } = props

  return (
    <>
      <button
        id={id}
        className={styles.button}
        onClick={() => setClicked(!clicked)}
      >
        {children}
      </button>
      <p className={styles.message}>
        {clicked ? 'Thanks for Clicking!' : 'Try Clicking..'}
      </p>
    </>
  )
}
