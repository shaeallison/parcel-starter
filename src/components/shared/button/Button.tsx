import * as styles from './button.module.css'
import classNames from 'classnames'
import { type ForwardedRef, forwardRef, type ReactNode } from 'react'
import { handleNewTab } from '../typography/Text'

/**
 * Button
 * @param {string} style - The type of the button, currently supports "primary", "secondary", "icon-only", "link" and "function".
 * @param {string} [type] - supports button, submit and reset
 * @param {string} [href] - string for href property
 * @param {boolean} [disabled] - supports button, submit and reset
 * @param {string} [ariaLabel] - string for labeling element, most common
 * @param {string} [ariaDescribedBy] - string with longer description of element, ex tooltip content
 * @param {string} [as] - The html element needed. "button", "div" and "a" supported.
 * @param {ReactNode} children - The text or fragment nested inside the element.
 */

export interface ButtonProps {
  children: ReactNode
  style: string
  type?: 'button' | 'submit' | 'reset'
  as: string
  id?: string
  href?: string
  linkNewTab?: boolean
  ariaLabel?: string
  ariaDescribedBy?: string
  disabled?: boolean
  onClick?: (e: React.MouseEvent<HTMLElement>) => void | Promise<void>
  className?: string
  title?: string
}

export const Button = forwardRef(function Button(props: ButtonProps, ref) {
  const {
    children,
    style,
    type,
    as,
    href,
    linkNewTab = false,
    ariaLabel,
    ariaDescribedBy,
    disabled,
    onClick,
    className,
    id,
    title
  } = props
  const classes = classNames(`${styles[style] as string}`, className)

  const externalLinkProps: React.HTMLProps<HTMLAnchorElement> = {}
  if (linkNewTab) {
    externalLinkProps.onClick = (event) => {
      event.preventDefault()
      handleNewTab(href)
    }
    externalLinkProps.rel = 'external nofollow nooopener'
  }

  switch (as) {
    case 'button':
      return (
        <button
          type={type}
          disabled={disabled}
          className={classes}
          aria-label={ariaLabel ?? ''}
          aria-describedby={ariaDescribedBy ?? ''}
          {...(disabled ? { 'aria-disabled': true } : {})}
          onClick={disabled ? undefined : onClick}
          id={id}
          ref={ref as ForwardedRef<HTMLButtonElement>}
          title={title}
        >
          {children}
        </button>
      )
    case 'a': {
      return (
        <a
          href={href}
          aria-label={ariaLabel ?? ''}
          aria-describedby={ariaDescribedBy ?? ''}
          className={classes}
          onClick={onClick}
          {...externalLinkProps}
          id={id}
          ref={ref as ForwardedRef<HTMLAnchorElement>}
        >
          {children}
        </a>
      )
    }
    default:
      // TODO: remove this case if not needed, we shouldn't have an interactive component using a div
      return (
        <div
          className={classes}
          aria-label={ariaLabel}
          aria-describedby={ariaDescribedBy ?? ''}
          id={id}
          ref={ref as ForwardedRef<HTMLDivElement>}
        >
          {children}
        </div>
      )
  }
})
