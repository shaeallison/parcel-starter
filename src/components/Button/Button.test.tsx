/**
 * @jest-environment jsdom
 */

import { act, create } from 'react-test-renderer'
import { render } from '@testing-library/react'
import { Button } from './Button'

it('click button, see message change', () => {
  const component = <Button id='test-id'>Click Me!</Button>
  let tree = create(component)
  const { getByText } = render(component)

  expect(tree).toMatchSnapshot()

  // Initial message
  expect(getByText('Try Clicking..'))

  // manually trigger click
  act(() => {
    getByText('Click Me!').click()
  })

  // Second message
  expect(getByText('Thanks for Clicking!'))
})
