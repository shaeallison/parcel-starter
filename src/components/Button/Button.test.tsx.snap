// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`click button, see message 1`] = `
[
  <button
    id="test-id"
    onClick={[Function]}
  >
    Click Me!
  </button>,
  <p>
    Try Clicking..
  </p>,
]
`;
